File: 103.二叉树的锯齿形层次遍历/zigzagLevelOrder.go
package problem0103

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	var result [][]int

	var dfs func(node *TreeNode, level int)
	dfs = func(node *TreeNode, level int) {
		if node == nil {
			return
		}
		if level == len(result) {
			result = append(result, []int{})
		}
		dfs(node.Right, level+1)
		dfs(node.Left, level+1)
		if level&1 == 1 {
			result[level] = append(result[level], node.Val)
		} else {
			result[level] = append([]int{node.Val}, result[level]...)
		}
	}
	dfs(root, 0)

	return result
}

File: 103.二叉树的锯齿形层次遍历/zigzagLevelOrder.js
// 层序遍历加一个反转标识
function zigzagLevelOrder(root) {
	if (!root) return []
	const stack = [root]
	const result = []
	let toRight = true

	while (stack.length) {
		result.push([])
		const len = stack.length
		for (let i = 0; i < len; i++) {
			const node = stack.shift()
			toRight
				? result[result.length - 1].push(node.val)
				: result[result.length - 1].unshift(node.val)
			node.left && stack.push(node.left)
			node.right && stack.push(node.right)
		}
		toRight = !toRight
	}
	return result
}